.github/container/jax_nsys/python/jax_nsys/jax_nsys/data_loaders.py

```diff
@@ -33,44 +34,34 @@ def is_communication(row):
-        # This is a range annotating a communication operation, i.e. NCCL kernel
-        # That kernel was active from thunk_row.ProjStartNs until thunk_row.ProjEndNs
-        # but during that time then other computation was going on. We want to
-        # find how much of the time did not overlap with other computation.
-        compute_df = thunk_df[
-            (
-                (thunk_df["ProjEndNs"] > comm_thunk.ProjStartNs)
-                & (thunk_df["ProjStartNs"] < comm_thunk.ProjEndNs)
-                & ~thunk_df["Communication"]
-            )
-        ]
-        # The computation kernels should all be serialised, but check that
-        for row1, row2 in itertools.pairwise(compute_df.itertuples()):
+    for _, thread_df in thunk_df.groupby("TID"):
+        # Update the projected duration of each communication kernel to only
+        # include the non-overlapped time
+        for comm_thunk in thread_df[thread_df["Communication"]].itertuples():
+            # This is a range annotating a communication operation, i.e. NCCL kernel
+            # That kernel was active from comm_thunk.ProjStartNs until comm_thunk.ProjEndNs
+            # but during that time then other computation was going on. We want to
+            # find how much of the time did not overlap with other computation.
+            compute_df = thread_df[
+                (
+                    (thread_df["ProjEndNs"] > comm_thunk.ProjStartNs)
+                    & (thread_df["ProjStartNs"] < comm_thunk.ProjEndNs)
+                    & ~thread_df["Communication"]
+                )
+            ]
+            # The computation kernels should all be serialised, but check that
             assert (
-                row2.ProjStartNs >= row1.ProjEndNs
-            ), f"{row2.Name} starts at {row2.ProjStartNs} before {row1.Name} ends at {row1.ProjEndNs}"
-        compute_time = sum(
-            min(row.ProjEndNs, comm_thunk.ProjEndNs)
-            - max(row.ProjStartNs, comm_thunk.ProjStartNs)
-            for row in compute_df.itertuples()
-        )
-        # Update the projected duration of communication kernels to just be the
-        # time that is not hidden.
-        unhidden_comm_time = comm_thunk.ProjDurNs - compute_time
-        thunk_df.loc[comm_thunk.Index, "ProjDurNs"] = unhidden_comm_time
-
-    # We assume that there is no compute-compute overlap for now; check that.
-    # TODO: be smarter about this if it's seen to take non-trivial time.
-    compute_df = thunk_df[~thunk_df["Communication"]]
-    for compute_thunk in compute_df.itertuples():
-        # This should just find the thunk itself
-        mask = (compute_df["ProjEndNs"] > compute_thunk.ProjStartNs) & (
-            compute_df["ProjStartNs"] < compute_thunk.ProjEndNs
-        )
-        assert mask.sum() == 1
+                compute_df["ProjStartNs"].iloc[1:].array
+                >= compute_df["ProjEndNs"].iloc[:-1].array
+            ).all()
```

Comment:
Is `compute_df` known to be sorted by `ProjStartNs`? (I assume so as otherwise this is unlikely to ever have succeeded?)

Reply:
I think we're allowed to expect that the data frame is sorted by launch time, which should mean it is also sorted by execution start time (under the hypothesis being asserted). I think it's conceivable that the assertion may fail in future, as I think in theory CUDA graphs could allow compute-compute overlap - but that failure is desirable.